README.md
# CARREFOUR DRIVE MONITOR
An helper to face COVID-19        
[FRENCH](#French)      
[ENGLISH](#English)    
[INSTALL](#INSTALL)     
# French
Le COVID-19 bouleverse nos habitudes et nous demande d'être plus prudent pour nous même et pour autrui.      
Faire ses courses aux Drive d'hypermarchés est essentiel pour des personnes âgées, des personnes possèdant un faible système immunitaire et généralement toute personne à risque.     
## Que permet ce monitor?
Il permet, à l'aide de l'API Carrefour, de détecter tous les points Drive disponibles aux alentours ou dans toute la France et de les ressencer.     
Deux modes sont pour le moment codés:     
- 1. Détection par code postal
- 2. Recensement de tous les Drive Carrefour de France et leur accessibilité actuelle     

# English
COVID-19 affects our routine and forces us to be more careful with ourselves as much as for others.     
Shop at supermarkets drive-in is essential for elderlies or people with a weak immunity system.     
## What's the goal of this monitor?
Thanks to Carrefour API, it can detect every drive-in available near your location or all around France and list them.     
Currently supporting two modes:     
- 1. Near drive-in by ZIP code     
- 2. Search for every drive-in with stock     

# INSTALL
## Requirements:    
- Python 3       
```pip3 install -r requirements.txt```
## Test:
- Rename the `config.ini.example` to `config.ini`
- Add your proxies below `[proxies]`     
- Add your postcode     
- `python3 main.py`     
### Options:
1. Monitor the stores near your location
2. Get stats about every Drive in France


CHANGELOG.md
# Changelog

## 0.1

**Implemented enhancements:**

- Monitor stores by postcode
- Get all the stores

**Fixed bugs:**



index.js
/**
 * Responds to any HTTP request.
 *
 * Redirect to handler depending on HTTP method.
 * @param {!express:Request} req HTTP request context.
 * @param {!express:Response} res HTTP response context.
 */
exports.processFbMessage = async (req, res) => {
	switch (req.method) {
	case 'GET':
		await handleGET(req, res);
		break;
	case 'POST':
		await handlePOST(req, res);
		break;
	default:
		console.error(`Unhandled http method ${req.method}`);
	}
};

/** Setting const and firebase requirements     */
const request = require('request');
const admin = require('firebase-admin');
const functions = require('firebase-functions');

admin.initializeApp(functions.config().firebase);

const db = admin.firestore();
/*******************/

/**
 * Handling GET to the server
 * @param {!express:Request} req HTTP request context.
 * @param {!express:Response} res HTTP response context.
 */
async function handleGET(req, res) {
	const mode = req.query['hub.mode'];
	const token = req.query['hub.verify_token'];
	const challenge = req.query['hub.challenge'];
	if (mode && token) {
		if (mode == 'subscribe' && token == process.env.VERIFY_TOKEN) {
			await res.status(200).send(challenge);
		} else {
			await res.sendStatus(403);
		}
	}else{
		await res.sendStatus(404).send("Not found :(");
	}
}

/**
 * Handling POST to the server
 * Exclusively reacting to Messenger "page" object.
 * Piling the different received messages to a Promise queue.
 * Once we receive confirmation that every message is treated we send back 200.
 * @param {!express:Request} req HTTP request context.
 * @param {!express:Response} res HTTP response context.
 */
async function handlePOST(req, res) {
	 // Parse the request body from the POST
	let body = req.body;

	// Check the webhook event is from a Page subscription
	if (body.object === 'page') {

	let promiseArray = body.entry.map(entry => {
		return new Promise((resolve) => {
			let webhook_event = entry.messaging[0];
			//console.log(webhook_event);
			console.log("sender: "+webhook_event.sender.id);
			console.log("timestamp: "+webhook_event.timestamp);
			console.log("recipient: "+webhook_event.recipient.id);
			console.log("message: "+webhook_event.message.text);
			resolve(handleMessage(webhook_event.sender.id,webhook_event.message, webhook_event.timestamp));
		});
	});
	Promise.all(promiseArray).then(() => {
		console.log('done')
		res.status(200).send('EVENT_RECEIVED');
	});
	// Return a '200 OK' response to all events

	} else {
	// Return a '404 Not Found' if event is not from a page subscription
	res.sendStatus(404);
	}
}

/**
 * Handling POST to the server
 * Exclusively reacting to Messenger "page" object.
 * Piling the different received messages to a Promise queue.
 * Once we receive confirmation that every message is treated we send back 200.
 * @param int sender_psid Facebook identifier for the sender.
 * @param {!messenger:Webhook.Message} Messenger object text message.
 * @param int timestamp sent at.
 */
async function handleMessage(sender_psid, received_message, timestamp) {
	let response;
	console.log("HANDLING");
	if(isInfos(received_message.text)){
		console.log("ASKING FOR INFOS");
		response = {
			"text": `Ok we're searching for your infos!`
		}
		console.log("SENDING")
		let a = callSendAPI(sender_psid, response);
		console.log("SENT")
		let users = db.collection('users');
		let snapshot = await users.where('psid', '==', sender_psid).get();
		if(snapshot.empty){
			console.log('No matching documents.');
			response = {
				"text": `We can't find you... 😢`
			}
			return callSendAPI(sender_psid, response);
		}
		for(doc of snapshot.docs){
			console.log(doc.id, '=>', doc.data());
			response = {
				"text": `Found you! Mister ${doc.id}`
			}
			return callSendAPI(sender_psid, response);
		}
	}
	
	else if (!(isPostCode(received_message.text))) {		
		// Create the payload for a basic text message
		response = {
			"text": `🤖 Je n'ai pas compris votre demande.\nMerci de saisir un code postal français valide.`
		}
		return callSendAPI(sender_psid, response);
	}
	else if(isPostCode(received_message.text)){
		let users = db.collection('users');
		let snapshot = await users.where('psid', '==', sender_psid).get();
		if(snapshot.empty){
			console.log("User not in database so we're adding him!")
			let doc = await addUser( sender_psid, timestamp, received_message.text.trim());
			console.log(`ID received by firestore: ${doc}`);
			response = {
				"text": `
					Merci! Vous êtes maintenant enregistré sous l'ID: "${doc}". \n
					Nous vous recontacterons si nous trouvons des disponibilités autour de chez vous! 😄\n
					Bon courage! 💪
				`
			}
			return callSendAPI(sender_psid, response);
		}else{
			for(doc of snapshot.docs){
				console.log("ALREADY INSIDE THE DB");
				// vous êtes ddéjà dans la bd
				response = {
					"text": `Vous êtes déjà dans la base de données!\nNous modifions donc votre code postal... 📝 `
				}
				let a = callSendAPI(sender_psid, response);
				// update avec un nouveau postcode
				let update = await users.doc(doc.id).update({postcode:received_message.text.trim()});
				response = {
					"text": `C'est bon!✅\nVotre code postal a bien été modifié.\nVous recevrez des alertes dès que des Drive autour de chez vous auront des disponibilités! 🚘`
				}
				return callSendAPI(sender_psid, response);
			}
			
		}
		
	}			
}

/**
 * Calling the Facebook API to send a message
 * @param int sender_psid Facebook identifier for the sender.
 * @param string response text to send.
 */
async function callSendAPI(sender_psid, response) {
	// Construct the message body
	let request_body = {
		"recipient": {
			"id": sender_psid
		},
		"message": response
	}
	request({
		"uri": "https://graph.facebook.com/v6.0/me/messages",
		"qs": { "access_token": process.env.PAGE_TOKEN },
		"method": "POST",
		"json": request_body
	}, (err, res, body) => {
		if (!err) {
			console.log('message sent!');
		} else {
			console.error("Unable to send message:" + err);
		}
	}); 
}

/*
 * Return if str is a valid french postcode
 * @param string str
 */
function isPostCode(str){
	return /^([0-9]{5})$/.test(str.trim());
}

/*
 * Return if str contains "info"
 * @param string str
 */
function isInfos(str){
	return str.trim().toLowerCase().match(/(info)/);
}

/**
 * Adding data to Cloud Firestore
 * @param int psid Facebook identifier for the sender.
 * @param int timestamp sent at.
 * @param string postcode given postcode.
 */
async function addUser(psid,timestamp,postcode){
	let addDoc = await db.collection('users').add({
		psid: psid,
		timestamp: timestamp,
		lastAsked: 0,
		lastPosted: 0,
		postcode: postcode 
	});
	console.log('Added document with ID: ', addDoc.id);
	return addDoc.id;
}

monitor.js
/**
 * Responds to any HTTP request.
 *
 * Redirect to handler depending on HTTP method.
 * @param {!express:Request} req HTTP request context.
 * @param {!express:Response} res HTTP response context.
 */
exports.processMonitor = async (req, res) => {
	switch (req.method) {
	case 'GET':
		await handleGET(req, res);
		break;
	default:
		console.error(`Unhandled http method ${req.method}`);
		await res.sendStatus(403);
	}
};

/** Setting const and firebase requirements */
const admin = require('firebase-admin');
const functions = require('firebase-functions');
const axios = require('axios');

admin.initializeApp(functions.config().firebase);

const db = admin.firestore();
/*******************/

async function handleGET(req,res){
	const mode = await req.query['mode'];
	const codepostal = await req.query['zip'];
	if(mode && codepostal){
		if(mode == "loca"){
			let carrefour = new Carrefour(codepostal);
			const location = await carrefour.getLoca();
			if(location.lat){
				let stores = await carrefour.getStores(location.lat,location.lng,codepostal)
				let availabilities = [];
				for(let store of stores){
					availabilities.push(await carrefour.getAvailability(store));
				}
				return new Promise(resolve=>{
					resolve(res.status(200).send(JSON.stringify(availabilities)))
				})
			}else{
				return new Promise(resolve=>{
					resolve(res.status(500).send(JSON.stringify({success:false,message:"Impossible d'obtenir la geolocalisation: "+location.message})))
				})
			}
			
		}
	}
	console.log("Not getting there")
	let zip = await monitor();
	await res.status(200).send(JSON.stringify(zip));
}

async function monitor(){
	let zip = [];
	let users = await db.collection('users').get();
	return users.docs.map(doc => doc.data().postcode);
}

class Carrefour{
	constructor(postcode){
		this.postcode = postcode;
		this.woosmap = "https://api.woosmap.com/localities/autocomplete/?key=woos-26fe76aa-ff24-3255-b25b-e1bde7b7a683&components=country:fr|country:mc&input="
		this.carrefourGeoloc = "https://www.carrefour.fr/geoloc"
		this.availability = "https://www.carrefour.fr/api/firstslot?storeId="
	}

	async getLoca(){
		return new Promise(resolve => {
			const headers = {
				'authority': 'api.woosmap.com',
				'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.69 Safari/537.36','accept': '*/*',
				'origin': 'https://www.carrefour.fr','sec-fetch-site': 'cross-site','sec-fetch-mode': 'cors','sec-fetch-dest': 'empty',
				'referer': 'https://www.carrefour.fr/services/drive','accept-language': 'en,en-US;q=0.9,fr;q=0.8',
			}
			axios.get(this.woosmap+this.postcode,{headers:headers})
				.then(response => {
					if(response.status === 200){
						if(response.data.localities[0]){
							resolve(response.data.localities[0].location) // = {"lat": 48.791378, "lng": 2.6660678}
						}else{
							resolve({"success":false,"message":"Le code postal ne retourne aucun résultat"})
						}
					}
					else{
						console.error("Seems like the request to get locals went wrong 1")
						console.error(response.status)
						console.error(response.data)
						resolve({"success":false,"message":"Impossible de joindre le serveur de récupération de localisation"})
					}
				})
				.catch(error => {
					console.error("Seems like the request to get locals went wrong 2")
					console.error(error)
					resolve({"success":false,"message":"Impossible de joindre le serveur de récupération de localisation","error":error})
				});
		});
	}

	async getStores(lat,lng,postcode){
		return new Promise(resolve => {
			console.log("Getting the stores")
			const headers = {
				'authority': 'www.carrefour.fr', 'accept': 'application/json, text/plain, */*',
				'x-requested-with': 'XMLHttpRequest', 'adrum': 'isAjax:true',
				'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.69 Safari/537.36',
				'sec-fetch-site': 'same-origin','sec-fetch-mode': 'cors','sec-fetch-dest': 'empty',
				'referer': 'https://www.carrefour.fr/services/drive','accept-language': 'en,en-US;q=0.9,fr;q=0.8'
			}
			const query = `?lat=${lat}&lng=${lng}&page=1&limit=5&postal_code=${postcode}&array_postal_codes%5B%5D=${postcode}&modes%5B%5D=delivery&modes%5B%5D=picking`
			axios.get(this.carrefourGeoloc+query,{headers:headers})
				.then(response => {
					if(response.data.data){
						const stores = response.data.data.stores;
						if(stores.length != 0){
							resolve(stores)
						}else{
							resolve({"success":false,"message":"Impossible de trouver un magasin autour de chez vous chez Carrefour."})
						}
					}
					else{
						if(response.data.contains("maintenance")){
							console.error("Carrefour est en maintenance pour le moment")
							console.error(response.status)
							resolve({"success":false,"message":"Carrefour est en maintenance pour le moment"})
						}else{
							console.error("Erreur inconnue, impossible de joindre Carrefour")
							console.error(response.status)
							resolve({"success":false,"message":"Erreur inconnue, impossible de joindre Carrefour"})
						}
					}
				})
				.catch(error => {
					console.error("Seems like the request to get stores went wrong")
					console.error(error)
					resolve({"success":false,"message":"Impossible de joindre Carrefour, impossible de trouver des magasins","error":error})
				});
		});
	}

	async getAvailability(store){
		return new Promise(resolve => {
			const headers = {
				'Host': 'www.carrefour.fr','accept': 'application/json, text/plain, */*','user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.69 Safari/537.36',
				'x-requested-with': 'XMLHttpRequest', // needed to get redirected to their Ajax Controller
				'sec-fetch-site': 'same-origin','sec-fetch-mode': 'cors','sec-fetch-dest': 'empty','referer': 'https://www.carrefour.fr/', 'accept-language': 'en,en-US;q=0.9,fr;q=0.8',
			}
			//console.log(store)
			axios.get(this.availability+store.ref,{headers:headers})
				.then(response => {
					console.log(response.status)
					if(response.data.data){
						if(response.data.data.type){
							resolve({"store":store,"availability":response.data.data.attributes.begDate,"error":false})
						}else{ 
							resolve({"store":store,"availability":null,"error":true,"message":response.data})
						}
					}
					else if(JSON.stringify(response.data) === '[]'){
						resolve({"store":store,"availability":null,"error":false})
					}
					else{
						if(JSON.stringify(response.data).includes("maintenance")){
							console.error("Carrefour est en maintenance pour le moment")
							console.error(response.status)
							resolve({"success":false,"message":"Carrefour est en maintenance pour le moment"})
						}else{
							console.error("Erreur inconnue, impossible de joindre Carrefour")
							console.error(response.status)
							console.error(response.data)
							resolve({"success":false,"message":"Erreur inconnue, impossible de joindre Carrefour"})
						}
					}
				})
				.catch(error => {
					console.error("Seems like the request to get stores went wrong")
					console.error(error)
					resolve({"success":false,"message":"Impossible de joindre Carrefour, impossible de trouver des magasins","error":error})
				});
		});
	}
}

main.py
import requests, random, datetime, json, threading, time, sys
from configparser import ConfigParser
configparser = ConfigParser(delimiters=('='),allow_no_value=True)

stores = []
storesErr = []
countAvail = 0
countNotAvail = 0

api = "https://www.carrefour.fr/api/firstslot?storeId="

headers = {
	'Host': 'www.carrefour.fr','accept': 'application/json, text/plain, */*','user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.69 Safari/537.36',
	'x-requested-with': 'XMLHttpRequest', # needed to get redirected to their Ajax Controller
	'sec-fetch-site': 'same-origin','sec-fetch-mode': 'cors','sec-fetch-dest': 'empty','referer': 'https://www.carrefour.fr/', 'accept-language': 'en,en-US;q=0.9,fr;q=0.8',
}


def log(event):
	d = datetime.datetime.now().strftime("%H:%M:%S")
	print("@azrpas :: " + str(d) + " :: " + event)

def grabProxies():
	configparser.read('config.ini')
	proxies = list(configparser.items('proxies'))
	log("Found "+str(len(proxies))+" proxies")
	return proxies

def getLoca(postcode):
	headers = {
		'authority': 'api.woosmap.com',
		'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.69 Safari/537.36','accept': '*/*',
		'origin': 'https://www.carrefour.fr','sec-fetch-site': 'cross-site','sec-fetch-mode': 'cors','sec-fetch-dest': 'empty',
		'referer': 'https://www.carrefour.fr/services/drive','accept-language': 'en,en-US;q=0.9,fr;q=0.8',
	}
	params = (
		('key', 'woos-26fe76aa-ff24-3255-b25b-e1bde7b7a683'),
		('input', postcode),
		('components', 'country:fr|country:mc'),
	)

	try:
		response = requests.get('https://api.woosmap.com/localities/autocomplete/', headers=headers, params=params, timeout=30)
	except Exception as e:
		log("Error while getting the local. Please check your internet access.")
		print(e)
		return False
	if(response.status_code == 200):
		return response.json()["localities"][0]["location"] # = {"lat": 48.791378, "lng": 2.6660678}
	else:
		log("Seems like the request to get locals went wrong")
		return False

def getStores(locations,postcode):
	headers = {
		'authority': 'www.carrefour.fr', 'accept': 'application/json, text/plain, */*',
		'x-requested-with': 'XMLHttpRequest', 'adrum': 'isAjax:true',
		'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.69 Safari/537.36',
		'sec-fetch-site': 'same-origin','sec-fetch-mode': 'cors','sec-fetch-dest': 'empty',
		'referer': 'https://www.carrefour.fr/services/drive','accept-language': 'en,en-US;q=0.9,fr;q=0.8'
	}
	params = (
		('lat', locations["lat"]),
		('lng', locations["lng"]),
		('page', '1'),
		('limit', '5'),
		('postal_code', postcode),
		('array_postal_codes[]', postcode),
		('modes[]', ['delivery', 'picking']),
	)
	try:
		response = requests.get('https://www.carrefour.fr/geoloc', headers=headers, params=params)
	except Exception as e:
		log("Error while getting the stores from Carrefour. Please check your internet access.")
		print(e)
		return False
	if(response.status_code == 200):
		try:
			response.json()
		except Exception as e:
			log("Can't read JSON")
			if("maintenance" in response.text.lower()):
				log("Carrefour is in maintenance, will be waiting an hour.")
				time.sleep(3600)
				getStores(locations,postcode)
			else:
				print(response.text)
			return False
		stores = response.json()["data"]["stores"]
		if(len(stores) == 0):
			log("Can't find any store... Please check your settings!")
			return False
		log("Found "+str(len(stores))+" stores")
		return stores
	else:
		log("Seems like the request to get locals went wrong")
		return False


def getStore(store,proxy):
	global stores
	global storesErr
	global countAvail
	global countNotAvail
	s = requests.session()
	try:
		r = s.get(api+str(store),headers=headers,proxies=proxy,timeout=10)
	except requests.exceptions.Timeout:
		log("Timeout, retrying...")
		return getStore(store,proxy)
	except requests.exceptions.ConnectionError:
		log("Connection error, retrying...")
		return getStore(store,proxy)
	except requests.exceptions.ProxyError:
		log("Connection error, retrying...")
		return getStore(store,proxy)
	except Exception as e:
		return {"store":None,"availability":None,"error":True,"message":repr(e)}
	try:
		r.json() # testing if json can be read
	except Exception as e:
		log("Can't read JSON")
		print(r.status_code)
		if("maintenance" in r.text.lower()):
			log("Carrefour is in maintenance, will be waiting an hour.")
			time.sleep(3600)
		else:
			print(r.text)
		return {"store":None,"availability":None,"error":True,"message":r.text}
	if(r.json() == []): # usually the response when a store have no stock
		print("Store but empty")
		return {"store":store,"availability":None,"error":False}
	elif(('data' in r.json())): 
		if(('type' in r.json()['data']) and (r.json()['data']['type'] == "first_slot")): ## found availability
			print("Store has availability")
			return {"store":store,"availability":r.json()['data']['attributes']['begDate'],"error":False}
		else: # unknown json
			print("ERROR 1")
			print(r.status_code)
			print(r.json())
			return {"store":store,"availability":None,"error":True,"message":r.text}
	else: # either not containing 'data' key or not '[]', unknown error for now
		print("ERROR 2")
		print(r.status_code)
		print(r.json())
		return None

def monitorStores(stores,proxy):
	for i in stores:
		result = getStore(i["ref"],proxy)
		if(result == None):
			continue
		elif(result["error"] == True):
			print(result["message"])
			continue
		elif(result["error"] == False):
			if(result["availability"] == None):
				log("No availability at "+i["name"])
				continue
			else:
				try:
					print("\n-------------------------------------------------")
					log("Found store with availability !")
					print(i["name"]+" à "+i["distance"]+"km")
					print(i["address"]["address1"]+" - "+i["address"]["city"])
					print(i["address"]["cityCode"])
					print("-------------------------------------------------\n")
				except Exception as e:
					log("Error while printing store")
					print(e)
				continue

def launchTask(rang,proxies,taskId):
	global stores
	global storesErr
	global countAvail
	global countNotAvail
	nbLoop = 1000
	lck = threading.Lock()
	for i in range(rang,rang+100):
		log(str(i)+"/"+str(nbLoop))
		pr = random.choice(proxies)[0] # [0] cause of tuple format given by config parser
		proxy = {"http":"http://"+pr,"https":"https://"+pr}
		log("THREAD "+str(taskId)+" : "+"Using proxy: "+proxy["http"])
		result = getStore(i,proxy)
		log("THREAD "+str(taskId)+" : "+"Sleeping 7 seconds before next request")
		time.sleep(7)
		lck.acquire()
		if(result == None):
			lck.release()
			continue
		elif(result["error"] == True):
			print(result["message"])
			storesErr.append({"store":result["store"],"availability":None})
			lck.release()
			continue
		elif(result["error"] == False):
			if(result["availability"] == None):
				log("THREAD "+str(taskId)+" : "+"Found store: No availability")
				countNotAvail += 1
				stores.append({"store":result["store"],"availability":None})
				lck.release()
				continue
			else:
				log("THREAD "+str(taskId)+" : "+"Found store: Availability")
				countAvail += 1
				stores.append({"store":result["store"],"availability":result["availability"]})
				lck.release()
				continue

def main():
	global stores
	global storesErr
	global countAvail
	global countNotAvail
	proxies = grabProxies()
	print("Carrefour Drive helper by @azrpas")
	print("1. To monitor your area (add your postcode to config.ini)")
	print("2. To get some stats about active Carrefour Drive")
	answer = input("> ").strip()
	if(int(answer) == int(1)):
		configparser.read('config.ini')
		postcode = configparser.get("options","postcode")
		result = getLoca(postcode)
		if(result == False):
			log("Can't get localisation, please check your parameters. Exiting.")
			sys.exit(1)
		result = getStores(result,postcode)
		if(result == False):
			log("Can't get stores after localisation, please check your parameters. Exiting.")
			sys.exit(1)
		while(True):
			pr = random.choice(proxies)[0]
			log("Checking your stores")
			proxy = {"http":"http://"+pr,"https":"https://"+pr}
			monitorStores(result,proxy)
			log("Sleeping a minute before checking again")
			time.sleep(60)
	elif(int(answer) == 2):
		rang = 0
		threads = []
		for i in range(0,10):
			log("Launching thread "+str(i))
			t = threading.Thread(target=launchTask, kwargs={"rang":rang,"proxies":proxies,"taskId":i})
			threads.append(t)
			t.start()
			rang += 100
		
		#save to file 
		for t in threads:
			t.join()
		log(str(countAvail)+" magasins ont leur fonctionnalité Carrefour Drive active")
		log(str(countNotAvail)+" magasins ont leur fonctionnalité Carrefour Drive désactivée")
		try:
			with open('stores.json', 'w') as f:
				toJson = {"stores":stores,"lastFetchedAt":datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")}
				json.dump(toJson, f)
			log("Successfully saved the stores 1/2")
		except Exception as e:
			log("Can't save them to JSON. Here's the error and the stores:")
			print(e)
			print(stores)
		try:
			with open('storesErr.json', 'w') as f:
				toJsonErr = {"stores":storesErr,"lastFetchedAt":datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")}
				json.dump(toJsonErr, f)
			log("Successfully saved the stores 2/2")
		except Exception as e:
			log("Can't save them to JSON. Here's the error and the stores:")
			print(e)
			print(storesErr)
	else:
		log("Please enter a proper value. Rebooting...")
		return main()
	
	
main()